### PR TITLE
fix(array): cast numeric string keys into `zend_ulong` and allow negative keys

### PR DESCRIPTION
**Pull request by kakserpom**
https://github.com/davidcole1340/ext-php-rs/issues/453

### CODE REVIEW

File: src/types/array.rs [hunk 1]
```diff
@@ -378,7 +379,11 @@ impl ZendHashTable {
         V: IntoZval,
```
Comment (Xenira) [inline src/types/array.rs:505]: Maybe add a generic `Into<ArrayKey>` here and work with that

**Review by Xenira — COMMENTED** [review]
Didn't have a detailed look yet, but remembered that we have an `ArrayKey` enum. Might want to utilise it here.

File: src/types/array.rs (hunk 1 above)
Comment (kakserpom) [reply src/types/array.rs:505]: ~~It would be great, but we can't use it  (`ArrayKey::Long(i64)`) until we fix https://github.com/davidcole1340/ext-php-rs/issues/454.~~

Comment (kakserpom) [reply src/types/array.rs:505]: I've added ArrayKeyRef.

**Comment by kakserpom** [comment]
PR solves both https://github.com/davidcole1340/ext-php-rs/issues/453 and https://github.com/davidcole1340/ext-php-rs/issues/454.

**Comment by Xenira** [comment]
@kakserpom Would you mind adding integration and unit tests?

**Comment by Xenira** [comment]
@kakserpom just a hint for commit messages you can use [convco](https://convco.github.io)

**Comment by kakserpom** [comment]
@Xenira Ordnung über alles? 😁

**Comment by Xenira** [comment]
@kakserpom keep it in english pls 😉

**Comment by kakserpom** [comment]
Squash the commits, please. @davidcole1340 @Xenira

**Comment by kakserpom** [comment]
@Qard done

**Review by Qard — COMMENTED** [review]
What about all the #[allow(clippy::mut_from_ref)] lines? Are those actually necessary?

**Comment by kakserpom** [comment]
@Qard yes, otherwise Clippy shows warnings because of &self instead of &mut self in function signatures.

**Comment by Qard** [comment]
Ah, so it's intentional to leave it then? Just wanted to be sure we weren't just forgetting to un-silence clippy in some places it might have legitimate warnings. 😅

**Comment by kakserpom** [comment]
Yes, It's intentional. I've enabled clippy::pedantic in another PR and quite a few warnings appeared, some of them led to actual fixes and some had to be silenced.

**Review by Qard — APPROVED** [review]
The clippy changes I guess aren't related to this PR, but necessary to get CI to pass, so that's fine.

Everything LGTM. Hopefully this can land soon to unblock other things. 🙂

**Comment by Xenira** [comment]
@kakserpom @Qard Regarding the clippy issues: Our usual procedure is to create a mr fixing those issues and rebase affected MRs. I have created one for the current issues, except the ones in #461, as those require some more investigation and possible breaking changes.

@kakserpom could you remove those ignores and rebase once #474 is merged?

File: src/types/array.rs [hunk 2]
```diff
@@ -25,6 +17,15 @@ use crate::{
...
 };
+use std::str::FromStr;
+use std::{
+    collections::HashMap,
+    convert::{TryFrom, TryInto},
+    ffi::CString,
+    fmt::{Debug, Display},
+    iter::FromIterator,
+    ptr,
+};
```
Comment (Xenira) [inline src/types/array.rs]: These can be merged

File: src/types/array.rs [hunk 3]
```diff
@@ -1,15 +1,7 @@
...
-use std::{
-    collections::HashMap,
-    convert::{TryFrom, TryInto},
-    ffi::CString,
-    fmt::{Debug, Display},
-    iter::FromIterator,
-    ptr,
-};
-
+use crate::ffi::zend_ulong;
```
Comment (Xenira) [inline src/types/array.rs]: This can be merged below.

We usually have the crate use statement below other ones. Maybe change this back.

File: src/types/array.rs [hunk 4]
```diff
@@ -606,15 +774,23 @@ pub struct Iter<'a> {
 }
 
 /// Represents the key of a PHP array, which can be either a long or a string.
-#[derive(Debug, PartialEq)]
-pub enum ArrayKey {
+#[derive(Debug, Clone, PartialEq)]
+pub enum ArrayKey<'a> {
```
Comment (Xenira) [inline src/types/array.rs:783]: Add a comment here to explain the unsigned cast. As the cast happens at multiple places it would be overkill to have it at all those places. Guess this is the best place for it. If you have a better place feel free to place it there.

File: crates/cli/src/lib.rs [hunk 5]
```diff
@@ -92,6 +92,7 @@ enum Args {
 
 #[allow(clippy::struct_excessive_bools)]
 #[derive(Parser)]
+#[allow(clippy::struct_excessive_bools)]
```
Comment (Xenira) [inline crates/cli/src/lib.rs]:
```suggestion
```

File: src/types/array.rs [hunk 6]
```diff
@@ -277,8 +356,12 @@ impl ZendHashTable {
     // hashtable while only having a reference to it. #461
     #[allow(clippy::mut_from_ref)]
     #[must_use]
-    pub fn get_index_mut(&self, key: u64) -> Option<&mut Zval> {
-        unsafe { zend_hash_index_find(self, key).as_mut() }
+    #[allow(clippy::mut_from_ref)]
```
Comment (Xenira) [inline src/types/array.rs]:
```suggestion
```

File: src/types/array.rs [hunk 7]
```diff
@@ -219,9 +257,47 @@ impl ZendHashTable {
     // hashtable while only having a reference to it. #461
     #[allow(clippy::mut_from_ref)]
     #[must_use]
-    pub fn get_mut(&self, key: &'_ str) -> Option<&mut Zval> {
-        let str = CString::new(key).ok()?;
-        unsafe { zend_hash_str_find(self, str.as_ptr(), key.len() as _).as_mut() }
+    #[allow(clippy::mut_from_ref)]
```
Comment (Xenira) [inline src/types/array.rs]:
```suggestion
```

File: src/types/array.rs (hunk 3 above)
Comment (kakserpom) [reply src/types/array.rs]: `cargo fmt` keeps doing this

Comment (Xenira) [reply src/types/array.rs]: interesting... let me check

**Comment by kakserpom** [comment]
@Xenira
> Could you also add integration tests for negative and 'string number' indices?

I'd rather someone else do it because I am on a trip until the end of the week. It's pretty trivial.

File: src/types/array.rs (hunk 3 above)
Comment (Xenira) [reply src/types/array.rs]: Ahhh, you were missing the empty line between them. Fixed it.

**Comment by Xenira** [comment]
> I'd rather someone do it because I am on a trip until the end of the week. It's pretty trivial.

Sure, will add them :)

**Review by Xenira — APPROVED** [review]
